scalding-macros/src/main/scala/com/twitter/scalding/macros/impl/ordered_serialization/ProductLike.scala

```diff
@@ -0,0 +1,186 @@
+          }
+      }
+
+    val combinedDynamic = dynamicFunctions.foldLeft(q"""$constSize""") {
+      case (prev, t) =>
+        q"$prev + $t"
+    }
+
+    if (noLength > 0) {
+      NoLengthCalculationAvailable(c)
+    } else {
+      if (maybeLength.size == 0 && dynamicFunctions.size == 0) {
+        ConstantLengthCalculation(c)(constSize)
+      } else {
+        if (maybeLength.size == 0) {
+          FastLengthCalculation(c)(combinedDynamic)
+        } else {
+
+          val const = q"_root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.ConstLen"
+          val dyn = q"_root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.DynamicLen"
+          val noLen = q"_root_.com.twitter.scalding.macros.impl.ordered_serialization.runtime_helpers.NoLengthCalculation"
+          // Contains an MaybeLength
+          val combinedMaybe: Tree = maybeLength.tail.foldLeft(maybeLength.head) {
+            case (hOpt, nxtOpt) =>
+              q"""
+              ($hOpt, $nxtOpt) match {
+                case ($const(l), $const(r)) => $const(l + r)
+                case ($const(l), $dyn(r)) => $dyn(l + r)
+                case ($dyn(l), $const(r)) => $dyn(l + r)
+                case ($dyn(l), $dyn(r)) => $dyn(l + r)
+                case _ => $noLen
+              }
+            """
+          }
+          if (dynamicFunctions.size > 0) {
+            MaybeLengthCalculation(c) (q"""
+            $combinedMaybe match {
+              case $const(l) => $dyn(l + $combinedDynamic)
+              case $dyn(l) => $dyn(l + $combinedDynamic)
+              case $noLen => $noLen
+            }
+          """)
+          } else {
+            MaybeLengthCalculation(c) (combinedMaybe)
+          }
+        }
+      }
+    }
+  }
+
+  def compare(c: Context)(elementA: c.TermName, elementB: c.TermName)(elementData: List[(c.universe.Type, c.universe.TermName, TreeOrderedBuf[c.type])]): c.Tree = {
+    import c.universe._
+
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val innerElementA = freshT("innerElementA")
+    val innerElementB = freshT("innerElementB")
+
+    val combinedData = elementData.map {
+      case (tpe, accessorSymbol, tBuf) =>
+        val curCmp = freshT("curCmp")
+        val curCmpFn = freshT("curCmpFn")
+        val cmpTree = q"""
+            val $curCmpFn: () => Int = () => {
+              val $innerElementA = $elementA.$accessorSymbol
+              val $innerElementB = $elementB.$accessorSymbol
+              ${tBuf.compare(innerElementA, innerElementB)}
+            }
+          """
+        (cmpTree, curCmpFn)
+    }
+    val fns = freshT("fns")
+    val goFn = freshT("goFn")
+    val tmpRes = freshT("goFn")
+    q"""
+      ..${combinedData.map(_._1)}
+      val $fns = List(..${combinedData.map(_._2)})
+      @_root_.scala.annotation.tailrec
+      def $goFn(l: List[() => Int]): Int = {
```

Comment:
We could, need a better pattern as how those things should fully unroll nicely. But yeah its effectively unrolled then re-rolled.